### PR TITLE
Fix for payment is rewritten

### PR DESCRIPTION
**Pull request by jonano614**

### CODE REVIEW

**Comment by sgaragagghu** [comment]
Seems fine :thumbsup:
I'll review when it is finished :D

**Comment by jonano614** [comment]
@xrdavies but where are those witness blocks generated?

File: client/block.c [hunk 1]
```diff
@@ -921,74 +923,97 @@ int xdag_create_block(struct xdag_field *fields, int inputsCount, int outputsCou
...
-		memcpy(task->lastfield.data, block[0].field[XDAG_BLOCK_FIELDS - 1].data, sizeof(struct xdag_field));
+	log_block("Create", newBlockHash, block[0].field[0].time, 1);
+	
+	if(block_hash_result != NULL) {
+		memcpy(block_hash_result, newBlockHash, sizeof(xdag_hash_t));
+	}
 
-		xdag_hash_final(task->ctx, &task->nonce.amount, sizeof(uint64_t), task->minhash.data);
-		g_xdag_pool_task_index = taskIndex;
+	struct xdag_block *new_block = (struct xdag_block *)malloc(sizeof(struct xdag_block));
```
Comment (sgaragagghu) [inline client/block.c]: Could you do an error check? :)

File: client/block.c [hunk 2]
```diff
@@ -921,74 +923,97 @@ int xdag_create_block(struct xdag_field *fields, int inputsCount, int outputsCou
...
+			pretop = pretop_new;
+			xdag_info("Mining: start from beginning because of pre-top block changed");
+			return 0;
 		}
-		res = 0;
 	}
 
-	return res;
-}
+	pthread_mutex_lock((pthread_mutex_t*)g_ptr_share_mutex);
```
Comment (sgaragagghu) [inline client/block.c]: where this lock is also used? i can't find it

Comment (xrdavies) [reply client/block.c]: It's a mutex for both miner and pool as far as I remember.

Comment (jonano614) [reply client/block.c]: This lock is not new, code section was just moved. It is necessary to protect the best found share between two threads.

File: client/block.c (hunk 1 above)
Comment (jonano614) [reply client/block.c]: Ok, but in my opinion we have to close the pool in case of insufficient memory.

**Comment by jonano614** [comment]
@xrdavies we need to discuss it.

File: client/block.c (hunk 2 above)
Comment (sgaragagghu) [reply client/block.c]: I know that it is not new, but i don't understand the logic of use a mutex to read task->lastfield.data while you write task->lastfield.data without this lock so i'm just asking what's the meaning of it
maybe i don't see where else this lock is used, i only see it here to read task->lastfield.data

File: client/block.c (hunk 1 above)
Comment (jonano614) [reply client/block.c]: Done

**Pull request merged**